sshkey-attest/src/lib.rs

```diff
@@ -0,0 +1,453 @@
+    })
+}
+
+macro_rules! cbor_try_bytes {
+    (
+        $v:expr
+    ) => {{
+        match $v {
+            serde_cbor::Value::Bytes(m) => Ok(m),
+            _ => Err(WebauthnError::COSEKeyInvalidCBORValue),
+        }
+    }};
+}
+
+#[derive(Debug)]
+struct SshSkAttestation {
+    att_cert: x509::X509,
+    sig: Vec<u8>,
+    auth_data_bytes: Vec<u8>,
+    auth_data: AuthenticatorData<Registration>,
+}
+
+struct SshSkAttestationRaw<'a> {
+    // This is the x5c cbor per https://developers.yubico.com/libfido2/Manuals/fido_cred_x5c_ptr.html
+    att_cert_raw: &'a [u8],
+    // Likely a cbor slice?
+    sig_raw: &'a [u8],
+    // cbor auth data. Could just be serde slice?
+    auth_data_raw: &'a [u8],
+}
+
+impl TryFrom<&[u8]> for SshSkAttestation {
+    type Error = WebauthnError;
+
+    fn try_from(data: &[u8]) -> Result<SshSkAttestation, WebauthnError> {
+        let sk_raw = ssh_sk_attestation_parser(data)
+            .map_err(|e| {
+                error!(?e, "try_from ssh_sk_attestation_parser");
+                WebauthnError::ParseNOMFailure
+            })
+            // Discard the remaining bytes.
+            .map(|(_, ad)| ad)?;
+
+        // Convert raw fields to parsed ones.
+
+        let sig = sk_raw.sig_raw.to_vec();
+
+        let att_cert =
+            x509::X509::from_der(sk_raw.att_cert_raw).map_err(WebauthnError::OpenSSLError)?;
+
+        let auth_data_bytes = serde_cbor::from_slice(sk_raw.auth_data_raw)
+            .map_err(|e| {
+                error!(?e, "invalid auth data cbor");
+                WebauthnError::ParseNOMFailure
+            })
+            .and_then(|value| cbor_try_bytes!(value))?;
+
+        let auth_data: AuthenticatorData<Registration> =
+            AuthenticatorData::try_from(auth_data_bytes.as_slice()).map_err(|e| {
+                error!(?e, "invalid auth data structure");
+                WebauthnError::ParseNOMFailure
+            })?;
+
+        Ok(SshSkAttestation {
+            att_cert,
+            sig,
+            // Probably need auth_data raw.
+            auth_data_bytes,
+            auth_data,
+        })
+    }
+}
+
+fn ssh_sk_attestation_parser(i: &[u8]) -> nom::IResult<&[u8], SshSkAttestationRaw> {
+    // Starts with a 4 byte u32 for the len of the header.
+
+    let (i, _tag_len) = tag([0, 0, 0, 17])(i)?;
+    let (i, _tag) = tag("ssh-sk-attest-v01")(i)?;
+
+    let (i, att_cert_len) = be_u32(i)?;
+    let (i, att_cert_raw) = take(att_cert_len as usize)(i)?;
+
+    let (i, sig_len) = be_u32(i)?;
+    let (i, sig_raw) = take(sig_len as usize)(i)?;
+
+    let (i, auth_data_len) = be_u32(i)?;
+    let (i, auth_data_raw) = take(auth_data_len as usize)(i)?;
+
+    let (i, _resvd_flags) = be_u32(i)?;
+    let (i, _resvd) = be_u32(i)?;
+
+    Ok((
+        i,
+        SshSkAttestationRaw {
+            att_cert_raw,
+            sig_raw,
+            auth_data_raw,
+        },
+    ))
+}
+
+fn to_ssh_pubkey(cose: COSEKey) -> Result<PublicKey, WebauthnError> {
```

Comment:
Could this be `impl TryFrom<CoseKey> for PublicKey`?

Reply:
Yes, but that then makes it public since you can't control visibility of impl trait.